// remote-modules/dist/server/installer/resource/adapters/css/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _parser = _interopRequireDefault(require("./parser"));

var _writer = _interopRequireDefault(require("../writer"));

var _default2 = _interopRequireDefault(require("../default"));

var _ = require("../");

var _escapeRegExp = _interopRequireDefault(require("../../../../../lib/helpers/escapeRegExp"));

var _requestAttributes = require("../../../../../lib/request-attributes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (C, ctx) => {
  const defaultAdapter = (0, _default2.default)(C, ctx);
  return (0, _.createAdapter)({
    outputType: 'css',
    parser: (0, _parser.default)(C),
    visitors: {
      Requests: {
        post: resource => {
          resource.requests.forEach(request => {
            const requestCtx = resource.contextFactory(request, resource.getResolverPaths());
            const {
              attributes,
              type,
              value
            } = request;

            switch (type) {
              case 'import':
                {
                  const nodes = resource.adapter.parser.runQuery(`
									// ${type} [ @import =~ '${(0, _escapeRegExp.default)(value)}' ]
								`);
                  nodes.forEach(node => {
                    if (requestCtx.isNormal()) {
                      // Drop import statements for normal resources
                      const {
                        rules
                      } = node.parent.stylesheet;
                      const index = rules.indexOf(node);
                      rules.splice(index, 1);
                    }
                  });
                  break;
                }

              default:
                {
                  const regexp = new RegExp(`(^|\\([\\'\\"]?)(?:<.+>)?${(0, _escapeRegExp.default)(value)}([\\'\\"]?\\)|$)`);
                  const nodes = resource.adapter.parser.runQuery(`
									// ${type} [ @value =~ '${regexp.source}' ]
								`);
                  nodes.forEach(node => {
                    const {
                      list
                    } = (0, _requestAttributes.parseAttributes)(node.value);

                    if (attributes === list || attributes && attributes.includes(list)) {
                      // eslint-disable-next-line no-param-reassign
                      node.value = node.value.replace(regexp, `$1${requestCtx.url}$2`);
                    }
                  });
                  break;
                }
            }
          });
        }
      },
      Complete: {
        post: resource => defaultAdapter.runVisitor(resource, ['Complete', 'post'])
      }
    },
    writer: new _writer.default({
      extension: '.css'
    })
  });
};

exports.default = _default;
