// remote-modules/dist/server/installer/resource/types/null.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _once = _interopRequireDefault(require("../../../../lib/helpers/once"));

var _normal = _interopRequireDefault(require("./normal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class NullResource extends _normal.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "getSourceChecksum", (0, _once.default)(() => this.sourceChecksum));

    _defineProperty(this, "getOptionsChecksum", (0, _once.default)(() => this.optionsChecksum));
  }

  getOutputSlug() {
    return this.slug;
  }

  traverse() {
    if (this.error) {
      this.logger.warn(`Dropped null resource ${this.moduleId}`);
    } else {
      this.logger.debug(`Skipped null resource ${this.moduleId}`);
    }

    this.dirty = false;
    this.loaded = true;
  } // eslint-disable-next-line class-methods-use-this


  getDeepDependencySet() {
    return new Set();
  }

}

exports.default = NullResource;
