// remote-modules/dist/server/installer/context/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isInstalledPackage = isInstalledPackage;
exports.getSlug = getSlug;
exports.getModuleId = getModuleId;
exports.getPackageId = getPackageId;
exports.getPid = getPid;
exports.getManifestPath = getManifestPath;
exports.slugToAbsolutePath = slugToAbsolutePath;
exports.default = ContextFactory;

var _path = _interopRequireDefault(require("path"));

var _url = _interopRequireDefault(require("url"));

var _util = require("util");

var _resolver = require("../../../lib/resolver");

var _helpers = require("../../../lib/helpers");

var _defineProperties = _interopRequireDefault(require("../../../lib/helpers/defineProperties"));

var _escapeRegExp = _interopRequireDefault(require("../../../lib/helpers/escapeRegExp"));

var _get = _interopRequireDefault(require("../../../lib/helpers/get"));

var _last = _interopRequireDefault(require("../../../lib/helpers/last"));

var _memoize = _interopRequireDefault(require("../../../lib/helpers/memoize"));

var _configStore = _interopRequireDefault(require("../../../lib/config-store"));

var _pid = _interopRequireDefault(require("../../../lib/helpers/pid"));

var _source = _interopRequireDefault(require("./types/source"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function isInstalledPackage(slug, moduleDirs) {
  return moduleDirs.some(dir => slug.includes(`${dir}/`));
}

function getSlug(origin, relativeTo) {
  return origin && (0, _helpers.isAbsolutePath)(origin) ? _path.default.relative(relativeTo, origin) : origin;
}

function getModuleId(slug, moduleDirs) {
  const isRelative = !(0, _resolver.isCore)(slug) && !moduleDirs.some(dir => slug.includes(`${dir}/`));

  const normalized = _path.default.relative('/', _path.default.normalize(`/${slug}`));

  const moduleDirsRegExp = new RegExp(`^(${moduleDirs.map(_escapeRegExp.default).join('|')})/`);
  return `${isRelative ? './' : ''}${normalized.replace(moduleDirsRegExp, '')}`;
}

function getPackageId(moduleId, moduleDirs) {
  if ((0, _helpers.isAbsoluteURL)(moduleId)) {
    return null;
  }

  const packageSegment = moduleDirs.reduce((result, dir) => (0, _last.default)(result.split(`${dir}/`)), moduleId);
  return packageSegment.match(/^(\.|(?:@[^/]+\/)?[^/]+)(?:\/|$)/)[1];
}

function getPid(moduleId) {
  return (0, _pid.default)(moduleId);
}

function getManifestPath(resolved) {
  return `${resolved}.manifest.js`;
}

function slugToAbsolutePath(baseDir, slug) {
  return _path.default.join(baseDir, _path.default.normalize(`/${slug}`));
}

function ContextFactory(options) {
  const C = _configStore.default.from(options);

  const cacheKeyResolver = (request, baseDir) => [(0, _get.default)(request, ['attributes']), (0, _get.default)(request, ['value']) || request, baseDir].join(':');

  const contextualResolve = (0, _memoize.default)((request, baseDir) => {
    let resolved;

    if (typeof request === 'string' && !(0, _helpers.isAbsolutePath)(request) && !(0, _helpers.isAbsoluteURL)(request)) {
      resolved = (0, _resolver.resolveSync)(_url.default.parse(request).pathname, _objectSpread({}, C.pick(['core', 'extensions', 'mainFields', 'moduleDirs', 'rootDir']), {
        baseDir: typeof baseDir === 'string' ? baseDir : C.rootDir
      }));
    } else {
      resolved = request;
    }

    return resolved;
  }, cacheKeyResolver);
  const createContext = (0, _memoize.default)((request, baseDir) => {
    // eslint-disable-next-line no-use-before-define
    const ctx = new _source.default(request, baseDir, factory, options);
    let iterations = 0;

    while (ctx.digest()) {
      C.runMiddleware('context', [ctx]);

      if (iterations > 1000) {
        throw new Error(`Exceeded max middleware iterations for '${ctx.moduleId}'`);
      }

      iterations += 1;
    }

    if (C.strict && ctx.error) {
      throw ctx.error;
    }

    return ctx;
  }, cacheKeyResolver);

  function factory(request, resolverPaths = [undefined]) {
    const requestObject = typeof request === 'string' ? {
      value: request
    } : request;

    if (!requestObject || !requestObject.value) {
      throw new Error(`Expected String or Object({ value: String }) but got '${(0, _util.inspect)(request)}'`);
    }

    if (!Array.isArray(resolverPaths)) {
      throw new Error(`Expected an array but got '${(0, _util.inspect)(resolverPaths)}'`);
    }

    let ctx;

    for (const baseDir of resolverPaths) {
      ctx = createContext(requestObject, baseDir);

      if (ctx.resolved) {
        break;
      }
    }

    if (ctx.error) {
      // Don't cache unresolvable requests
      factory.uncache({
        pid: ctx.pid
      });
    }

    return ctx;
  }

  return (0, _defineProperties.default)(factory, {
    name: 'ContextFactory',
    resolve: contextualResolve,

    safeResolve(...args) {
      try {
        return contextualResolve(...args);
      } catch (_err) {
        return null;
      }
    },

    getPid,

    getSlug(origin) {
      return getSlug(origin, C.rootDir);
    },

    getModuleId(slug, moduleDirs = C.moduleDirs) {
      return getModuleId(slug, moduleDirs);
    },

    getPackageId(moduleId, moduleDirs = C.moduleDirs) {
      return getPackageId(moduleId, moduleDirs);
    },

    isInstalledPackage(slug, moduleDirs = C.moduleDirs) {
      return isInstalledPackage(slug, moduleDirs);
    },

    uncache(query) {
      if (query) {
        for (const [key, ctx] of createContext.cache) {
          if ((0, _helpers.matches)(ctx, query)) {
            contextualResolve.cache.delete(key);
            createContext.cache.delete(key);
          }
        }
      } else {
        contextualResolve.cache.clear();
        createContext.cache.clear();
      }
    }

  });
}
