// remote-modules/dist/lib/logger/formatted-stream.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FormattedStream;

var _util = require("util");

var _safe = _interopRequireDefault(require("colors/safe"));

var _escapeRegExp = _interopRequireDefault(require("../helpers/escapeRegExp"));

var _2 = require("./");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

_safe.default.setTheme({
  trace: 'magenta',
  debug: 'cyan',
  info: 'green',
  warn: 'yellow',
  error: 'red',
  fatal: 'rainbow'
});

function formatSrc({
  file,
  line
}) {
  const pattern = new RegExp(`^${(0, _escapeRegExp.default)(process.cwd())}/(node_modules/)?`);
  const location = file.replace(pattern, '');
  return _safe.default.gray(`${location}:${line}`);
}

function format(_ref) {
  let {
    err,
    level,
    msg,
    name,
    src,
    time
  } = _ref,
      other = _objectWithoutProperties(_ref, ["err", "level", "msg", "name", "src", "time"]);

  const levelName = _2.Logger.nameFromLevel[level];
  return [time.toISOString(), `${_safe.default[levelName](levelName)}:`, `[${name}${src ? ` ${formatSrc(src)}` : ''}]`, err ? err.stack || err.message || err : null, err ? null : msg, !msg && !err ? `\n${(0, _util.inspect)(other, {
    colors: true
  })}` : null].filter(Boolean).join(' ');
}

function write(_ref2) {
  let {
    v: _
  } = _ref2,
      rec = _objectWithoutProperties(_ref2, ["v"]);

  switch (true) {
    case rec.level === _2.Logger.WARN:
      // eslint-disable-next-line no-console
      console.warn(format(rec));
      break;

    case rec.level >= _2.Logger.ERROR:
      // eslint-disable-next-line no-console
      console.error(format(rec));
      break;

    default:
      // eslint-disable-next-line no-console
      console.log(format(rec));
      break;
  }
}

function FormattedStream() {
  return {
    write
  };
}
