// remote-modules/dist/lib/logger/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Logger = void 0;

var _bunyan = _interopRequireDefault(require("bunyan"));

var _formattedStream = _interopRequireDefault(require("./formatted-stream"));

var _package = _interopRequireDefault(require("../../../package.json"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Logger extends _bunyan.default {
  constructor(options) {
    super(_objectSpread({
      name: _package.default.name,
      level: process.env.LOG_LVL,
      streams: [{
        type: 'raw',
        stream: new _formattedStream.default()
      }]
    }, options));
  }

  child(options) {
    return new this.constructor(_objectSpread({
      level: this.level(),
      streams: [...this.streams]
    }, options));
  }

  profile() {
    const start = process.hrtime();
    return label => {
      const [s, ns] = process.hrtime(start);
      const ms = s * 1e3 + ns * 1e-6;
      const message = label.message || label;
      const level = label.level || 'info';
      this[level](`${message} duration=${ms.toFixed(3)}ms`);
    };
  }

}

exports.Logger = Logger;

var _default = new Logger();

exports.default = _default;
