// remote-modules/dist/server/installer/resource/adapters/sass/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _deepmerge = _interopRequireDefault(require("deepmerge"));

var _nodeSass = _interopRequireDefault(require("node-sass"));

var _css = _interopRequireDefault(require("../css"));

var _promisify = _interopRequireDefault(require("../../../../../lib/helpers/promisify"));

var _fs = require("../../../../../lib/helpers/fs");

var _requestAttributes = require("../../../../../lib/request-attributes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const renderAsync = (0, _promisify.default)(_nodeSass.default.render);

var _default = (C, ctx) => {
  const adapter = (0, _css.default)(C, ctx);
  return _objectSpread({}, adapter, {
    visitors: (0, _deepmerge.default)(adapter.visitors, {
      Parse: {
        pre: async resource => {
          const {
            parser
          } = resource.adapter; // node-sass won't produce source maps if source is passed as
          // options.data and options.sourceMap is not a string, so we
          // just pass an arbitrary path relative to the module root.

          const sourceMap = resource.options.sourceMaps && _path.default.join(resource.options.rootDir, 'sass.map');

          const {
            css,
            map
          } = await renderAsync({
            file: resource.origin,
            data: resource.source,
            importer: (request, prev, done) => {
              const dependency = resource.resourceFactory((0, _requestAttributes.stripAttributes)(request), resource);
              const {
                attributes
              } = (0, _requestAttributes.parseAttributes)(request);

              if (attributes.href) {
                done({
                  contents: `@import url('${request}');`,
                  file: dependency.origin
                });
              } else if (dependency.isNormal()) {
                resource.resolverPaths.add(dependency.getOriginDir());
                Promise.resolve(dependency.source || (0, _fs.readFileAsync)(dependency.origin)).then(contents => {
                  done({
                    contents,
                    file: dependency.origin
                  });
                }).catch(done);
              } else {
                done(null);
              }
            },
            indentedSyntax: _path.default.extname(resource.slug) === '.sass',
            omitSourceMapUrl: true,
            sourceMap,
            sourceMapContents: sourceMap,
            sourceMapRoot: parser.sourceRoot
          });

          if (map) {
            const sourceMapJSON = _objectSpread({}, JSON.parse(map), {
              file: resource.slug
            });

            parser.sourceMapJSON = sourceMapJSON;
          } // eslint-disable-next-line no-param-reassign


          resource.output = css.toString('utf8');
          return adapter.runVisitor(resource, ['Parse', 'pre']);
        }
      }
    })
  });
};

exports.default = _default;
