// remote-modules/dist/lib/config-store/presets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _escapeRegExp = _interopRequireDefault(require("../helpers/escapeRegExp"));

var _helpers = require("../helpers");

var _resolver = require("../resolver");

var _middleware = require("./middleware");

var _ = _interopRequireDefault(require("./"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = preset => {
  switch (preset) {
    case 'node':
      return {
        define: {
          'process.browser': false,
          'typeof window': 'undefined',
          'typeof document': 'undefined',
          'typeof XMLHttpRequest': 'undefined'
        },
        sourceMaps: 'inline'
      };

    case 'browser':
      return {
        core: (0, _helpers.mapObject)(_resolver.defaultCore, () => false),
        mainFields: ['browser'],
        define: {
          'process.browser': true,
          'typeof window': 'object',
          'typeof document': 'object',
          'typeof XMLHttpRequest': 'function'
        },
        provide: {
          global: 'window',
          process: `import process from '${_.default.shims.process}'`,
          Buffer: `import { Buffer } from '${_.default.mocks.buffer}'`
        },
        middleware: [// Rewrite requests for node core modules to browser shims
        (0, _middleware.RewriteMiddleware)(new RegExp(`^(${Object.keys(_.default.shims).map(_escapeRegExp.default).join('|')})$`), ctx => _.default.shims[ctx.request])]
      };

    default:
      return {};
  }
};

exports.default = _default;
