// remote-modules/dist/server/installer/resource/adapters/less/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _deepmerge = _interopRequireDefault(require("deepmerge"));

var _less = _interopRequireDefault(require("less"));

var _css = _interopRequireDefault(require("../css"));

var _helpers = require("../../../../../lib/helpers");

var _fs = require("../../../../../lib/helpers/fs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ImportManager extends _less.default.FileManager {
  constructor(resource) {
    super();
    this.resource = resource;
  } // eslint-disable-next-line class-methods-use-this


  supports(filename) {
    return !(0, _helpers.isAbsoluteURL)(filename);
  } // eslint-disable-next-line class-methods-use-this


  supportsSync() {
    return false;
  }

  async loadFile(request) {
    const {
      resource
    } = this;
    const dependency = resource.resourceFactory(request, resource);

    if (dependency.error) {
      throw dependency.error;
    }

    resource.resolverPaths.add(dependency.getOriginDir());
    return {
      contents: dependency.source || (await (0, _fs.readFileAsync)(dependency.origin)),
      filename: dependency.slug
    };
  }

}

var _default = (C, ctx) => {
  const adapter = (0, _css.default)(C, ctx);
  return _objectSpread({}, adapter, {
    visitors: (0, _deepmerge.default)(adapter.visitors, {
      Parse: {
        pre: async resource => {
          const {
            parser
          } = resource.adapter;
          const {
            css,
            map
          } = await _less.default.render(resource.output, {
            filename: resource.slug,
            javascriptEnabled: true,
            sourceMap: Boolean(resource.options.sourceMaps) && {
              outputSourceFiles: true
            },
            relativeUrls: true,
            plugins: [{
              install(_, pluginManager) {
                pluginManager.addFileManager(new ImportManager(resource));
              },

              minVersion: [2, 3, 1]
            }]
          });

          if (map) {
            const sourceMapJSON = _objectSpread({}, JSON.parse(map), {
              sourceRoot: parser.sourceRoot
            });

            parser.sourceMapJSON = sourceMapJSON;
          } // eslint-disable-next-line no-param-reassign


          resource.output = css;
          return adapter.runVisitor(resource, ['Parse', 'pre']);
        }
      }
    })
  });
};

exports.default = _default;
