// remote-modules/dist/lib/manifest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Manifest {
  static load(internals) {
    return new Manifest(internals.dependencies, internals);
  }

  static derive(parent, meta) {
    const internals = parent.internals();
    return new Manifest(internals.dependencies, _objectSpread({}, internals, {
      meta
    }));
  }

  constructor(dependencies, internals = {}) {
    Object.defineProperty(this, Manifest.internals, {
      value: Object.freeze(_objectSpread({}, internals, {
        dependencies
      })),
      writable: false,
      enumerable: false
    });
    const lookupTable = new Map();
    const meta = this.meta();
    dependencies.forEach(value => {
      lookupTable.set(value.pid, value);
      lookupTable.set(value.moduleId, value);
    });
    lookupTable.set(meta.pid, meta);
    lookupTable.set(meta.moduleId, meta);
    this.lookupTable = lookupTable;
  }

  lookup(key) {
    return this.lookupTable.get(key);
  }

  exists(key) {
    return this.lookupTable.has(key);
  }

  getModuleId(key) {
    return this.exists(key) && this.lookup(key).moduleId || undefined;
  }

  getPid(key) {
    return this.exists(key) && this.lookup(key).pid || undefined;
  }

  getType(key) {
    return this.exists(key) && this.lookup(key).type || undefined;
  }

  getAssetId(key) {
    return this.exists(key) && this.lookup(key).assetId || undefined;
  }

  internals(key) {
    const internals = this[Manifest.internals];
    return key ? internals[key] : internals;
  }

  meta(key) {
    const meta = this.internals('meta');
    return key ? meta[key] : meta;
  }

  list() {
    return this.internals('dependencies').map(({
      moduleId
    }) => moduleId);
  }

  package(key) {
    return this.internals('packages')[key];
  }

  toJSON() {
    return this.internals();
  }

}

exports.default = Manifest;

_defineProperty(Manifest, "internals", Symbol('internals'));
