// remote-modules/dist/lib/request-attributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatAttributes = formatAttributes;
exports.parseAttributes = parseAttributes;
exports.stripAttributes = stripAttributes;

var _helpers = require("./helpers");

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function getAttributesRegExp() {
  return /(?:<([^>]+)>)/;
}

function formatAttributes(attributes) {
  const keys = Object.keys((0, _helpers.pickBy)(attributes, Boolean));
  return keys.length ? `<${keys.join(',')}>` : undefined;
}

function parseAttributes(requestValue) {
  const attributes = {};
  const [, attributesList] = requestValue.match(getAttributesRegExp()) || [];

  if (attributesList) {
    attributesList.split(',').forEach(attribute => {
      const [, key, value = true] = attribute.match(/^([^=]+)(?:=(.+)|$)/);
      attributes[key] = value;
    });
  }

  return {
    list: attributesList,
    attributes: _objectSpread({}, attributes, {
      href: attributes.href === undefined && attributes.static || attributes.href
    })
  };
}

function stripAttributes(requestValue) {
  return requestValue.replace(getAttributesRegExp(), '');
}
