// remote-modules/dist/client/registry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Registry {
  constructor(ttl) {
    _defineProperty(this, "expires", null);

    _defineProperty(this, "cache", new Map());

    _defineProperty(this, "links", new Map());

    if (typeof ttl !== 'number') {
      throw new Error('ttl is required');
    }

    this.ttl = ttl;
  }

  hasLink(key) {
    return this.links.has(key);
  }

  ln(alias, key) {
    this.links.set(alias, key);
  }

  lookup(key) {
    const {
      links
    } = this;
    return links.has(key) ? links.get(key) : key;
  }

  has(key) {
    return this.cache.has(this.lookup(key));
  }

  get(key) {
    return this.cache.get(this.lookup(key));
  }

  set(key, module) {
    const {
      cache
    } = this;
    const pid = this.lookup(key);

    if (!cache.has(pid)) {
      cache.set(pid, module);
    }
  }

  forEach(fn) {
    return this.cache.forEach(fn);
  }

  sweep() {
    if (!this.expires) {
      this.expires = Date.now() + this.ttl;
    }

    if (this.expires <= Date.now()) {
      this.clear();
    }
  }

  clear() {
    this.cache.clear();
    this.links.clear();
    this.expires = null;
  }

}

exports.default = Registry;
