// remote-modules/dist/client/module-transport/node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _noop = _interopRequireDefault(require("../../lib/helpers/noop"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ModuleTransport {
  constructor(loader) {
    _defineProperty(this, "pending", new Map());

    this.loader = loader;
  }

  // eslint-disable-next-line class-methods-use-this
  getResetPredicate() {// noop
  }

  getManifestJSON(id) {
    return this.loader.fetchManifestJSON(id);
  }

  initialize(request, parent) {
    const {
      loader,
      pending
    } = this;

    if (!pending.has(request)) {
      pending.set(request, new Promise(async (resolve, reject) => {
        try {
          let pid = loader.resolvePid(request, parent);
          let moduleId = loader.resolve(request, parent);
          let module;

          if (loader.getFromContext(pid)) {
            module = loader.register({
              content: 'this;',
              id: moduleId,
              pid,
              parent
            });
          } else {
            const url = loader.resolveURL(request, parent);
            const res = await loader.fetch(url);

            if (res.ok) {
              moduleId = res.headers['x-module-id'];
              pid = Number(res.headers['x-pointer-id']);
              module = loader.register({
                content: await res.text(),
                id: moduleId,
                pid,
                parent
              });
            } else {
              const _ref = await res.json(),
                    {
                message
              } = _ref,
                    other = _objectWithoutProperties(_ref, ["message"]);

              const messageWithContext = parent ? `${message}${parent.trace()}` : message;
              throw Object.assign(new Error(messageWithContext), other);
            }
          }

          resolve(module);
        } catch (err) {
          reject(err);
        }
      }));
      pending.get(request).catch(_noop.default).then(() => {
        pending.delete(request);
      });
    }

    return pending.get(request);
  }

}

exports.default = ModuleTransport;
