// remote-modules/dist/lib/helpers/fs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readJSONAsync = readJSONAsync;
exports.writeJSONAsync = writeJSONAsync;
exports.writeFileAsync = exports.readFileAsync = exports.statAsync = exports.mkdirpAsync = exports.rimrafAsync = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _mkdirp = _interopRequireDefault(require("mkdirp"));

var _rimraf = _interopRequireDefault(require("rimraf"));

var _promisify = _interopRequireDefault(require("./promisify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const rimrafAsync = (0, _promisify.default)(_rimraf.default);
exports.rimrafAsync = rimrafAsync;
const mkdirpAsync = (0, _promisify.default)(_mkdirp.default);
exports.mkdirpAsync = mkdirpAsync;
const statAsync = (0, _promisify.default)(_fs.default.stat, _fs.default);
exports.statAsync = statAsync;
const readFileAsync = (0, _promisify.default)(_fs.default.readFile, _fs.default, {
  defaults: {
    1: 'utf8'
  }
});
exports.readFileAsync = readFileAsync;
const writeFileAsync = (0, _promisify.default)(_fs.default.writeFile, _fs.default);
exports.writeFileAsync = writeFileAsync;

function readJSONAsync(path) {
  return readFileAsync(path).then(JSON.parse);
}

function writeJSONAsync(path, data, opts) {
  return writeFileAsync(path, JSON.stringify(data, null, 2), opts);
}
