// remote-modules/dist/lib/helpers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.matches = matches;
exports.isPrimitive = isPrimitive;
exports.isPlainObject = isPlainObject;
exports.mapObject = mapObject;
exports.pickBy = pickBy;
exports.pickDefined = pickDefined;
exports.omit = omit;
exports.stripBounding = stripBounding;
exports.isAbsoluteURL = isAbsoluteURL;
exports.isDataURL = isDataURL;
exports.isAbsolutePath = isAbsolutePath;
exports.isRelativePath = isRelativePath;

function matches(target, query) {
  let result = true;

  for (const [key, value] of Object.entries(query)) {
    if (target[key] !== value) {
      result = false;
      break;
    }
  }

  return result;
}

function isPrimitive(value) {
  return value === null || typeof value !== 'object' && typeof value !== 'function';
}

function isPlainObject(value) {
  return Boolean(value && (Object(value).constructor === Object || Object.getPrototypeOf(value) === null));
}

function mapObject(target, fn) {
  return target && Object.entries(target).reduce((acc, [key, value]) => {
    acc[key] = fn(value, key);
    return acc;
  }, Array.isArray(target) ? [] : {});
}

function pickBy(target, fn) {
  return target && Object.entries(target).reduce((acc, [key, value]) => {
    if (fn(value, key)) {
      acc[key] = value;
    }

    return acc;
  }, Array.isArray(target) ? [] : {});
}

function pickDefined(target) {
  return pickBy(target, value => value !== undefined);
}

function omit(target, keys) {
  return pickBy(target, (value, key) => !keys.includes(key));
}

function stripBounding(value, pattern) {
  return value && value.replace(new RegExp(`^${pattern}|${pattern}$`), '');
}

function isAbsoluteURL(path) {
  return Boolean(path && /^(\w+:)?\/\//.test(path));
}

function isDataURL(path) {
  return Boolean(path && /^data:\w+/i.test(path));
}

function isAbsolutePath(path) {
  return Boolean(path && /^\/([^/]|$)/.test(path));
}

function isRelativePath(path) {
  return Boolean(path && /^\.{1,2}(?:\/|$)/.test(path));
}
